### PR TITLE
[Feature][Clickhouse] Support sink savemode 

### LINKED ISSUE #8060: [Feature][Clickhouse] Support sink savemode
### Search before asking

- [X] I had searched in the [feature](https://github.com/apache/seatunnel/issues?q=is%3Aissue+label%3A%22Feature%22) and found no similar feature requirement.

### Description

[Feature][Clickhouse] Support sink savemode

reference
https://github.com/apache/seatunnel/blob/dev/seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/sink/JdbcSink.java#L65

updates
- update connector
- add e2e
- update docs 

### Usage Scenario

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by zhangshenghang**
### Purpose of this pull request

close #8060

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  2. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
  3. Add ci label in [label-scope-conf](https://github.com/apache/seatunnel/blob/dev/.github/workflows/labeler/label-scope-conf.yml)
  4. Add e2e testcase in [seatunnel-e2e](https://github.com/apache/seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/)
  5. Update connector [plugin_config](https://github.com/apache/seatunnel/blob/dev/config/plugin_config)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/catalog/ClickhouseCatalog.java [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    }
+
+    @SuppressWarnings("MagicNumber")
+    private Map<String, String> buildConnectorOptions(TablePath tablePath) {
+        Map<String, String> options = new HashMap<>(8);
+        options.put("connector", "clickhouse");
+        options.put("host", readonlyConfig.get(HOST));
+        options.put("database", tablePath.getDatabaseName());
+        options.put("username", readonlyConfig.get(USERNAME));
+        options.put("password", readonlyConfig.get(PASSWORD));
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/catalog/ClickhouseCatalog.java]: please do not put username/password into options.

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/sink/client/ClickhouseSinkWriter.java [hunk 2]
```diff
@@ -63,6 +63,7 @@ public class ClickhouseSinkWriter
         this.proxy = new ClickhouseProxy(option.getShardMetadata().getDefaultShard().getNode());
         this.shardRouter = new ShardRouter(proxy, option.getShardMetadata());
         this.statementMap = initStatementMap();
+        System.out.println(123);
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/sink/client/ClickhouseSinkWriter.java]: revert.

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/util/ClickhouseCatalogUtil.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                            .collect(Collectors.joining(","));
+        }
+        SqlTemplate.canHandledByTemplateWithPlaceholder(
+                template,
+                SaveModePlaceHolder.ROWTYPE_PRIMARY_KEY.getPlaceHolder(),
+                primaryKey,
+                TablePath.of(database, table).getFullName(),
+                ClickhouseConfig.SAVE_MODE_CREATE_TEMPLATE.key());
+        template =
+                template.replaceAll(
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/util/ClickhouseCatalogUtil.java]: how about put template parse into common module? Not only one connector use same logic.

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/util/CreateTableParser.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+import lombok.Getter;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class CreateTableParser {
+
+    private static final Pattern COLUMN_PATTERN = Pattern.compile("`?(\\w+)`?\\s*([\\w|\\W]*)");
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/util/CreateTableParser.java]: ditto

**Review by Hisoka-X — COMMENTED** [review]
Thanks @zhangshenghang ! Looks great!

File: seatunnel-connectors-v2/connector-clickhouse/pom.xml [hunk 5]
```diff
@@ -68,6 +68,11 @@
             <version>${project.version}</version>
             <classifier>optional</classifier>
         </dependency>
+        <dependency>
+            <groupId>org.apache.seatunnel</groupId>
+            <artifactId>connector-common</artifactId>
+            <version>${project.version}</version>
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-clickhouse/pom.xml:74]: please change the scope to provide. We should not put the common class into connector jar.

Comment (zhangshenghang) [reply seatunnel-connectors-v2/connector-clickhouse/pom.xml:74]: I have modified it, but I found that the scope of other connectors is compile, not is provided.

image

File: docs/en/connector-v2/sink/Clickhouse.md [hunk 6]
```diff
@@ -59,8 +59,69 @@ They can be downloaded via install-plugin.sh or from the Maven central repositor
...
+ORDER BY (${rowtype_primary_key})
+PRIMARY KEY (${rowtype_primary_key})
+SETTINGS
+    index_granularity = 8192;
+```
+
+If custom fields are added to the template, for example, adding an `id` field:
+
+```sql
+CREATE TABLE IF NOT EXISTS `${database}`.`${table_name}` (
```
Comment (hailin0) [inline docs/en/connector-v2/sink/Clickhouse.md]: Use `${table}`?

Comment (zhangshenghang) [reply docs/en/connector-v2/sink/Clickhouse.md]: > Use `${table}`?

@hailin0 already fixed

**Review by Hisoka-X — DISMISSED** [review]
LGTM. Thanks @zhangshenghang !

**Comment by Hisoka-X** [comment]
Please resolve conflicts.

**Review by Hisoka-X — DISMISSED** [review]

**Review by Hisoka-X — APPROVED** [review]

**Review by hailin0 — APPROVED** [review]

**Pull request merged**